docs/design/multipart-upload.md

```diff
@@ -0,0 +1,82 @@
+# Multipart Upload
+
+## Abstract
+
+This design document describes multipart upload and how it can be implemented.
+Currently, there is no direct support for multipart upload on the Satellite side.
+The feature is simulated on the gateway side, which requires keeping 
+a lot of data in memory and doesn't achieve the full usefulness of multipart upload.
+
+## Background
+
+Multipart upload gives quite a few nice properties:
+
+* Concurrent uploads,
+* continuation of partial uploads, and
+* out-of-order uploads.
+
+The flow for a multipart upload, from a S3 perspective, is:
+
+1. starts a new multipart upload,
+2. the upload data is split into multiple parts,
+3. for each part: send the part,
+4. calls "complete multipart upload," or
+5. alternatively, calls "cancel multipart upload."
+
+There are a few important things we must keep in mind:
+
+* the parts can have different sizes,
+* during the start of the upload, the part size can be unknown,
+* the parts can be uploaded from different computers,
+* the parts can be uploaded out of order, and
+* the parts can be reuploaded, in case of a failed upload.
+
+## Design
+
+We change the path handling to support listing parts uploaded in arbitrary order. When we start an upload we will create an object at the following path:
+
+```
+<project-id>/<bucket-id>/objects/<path ...>_<version> => object information (partial)
+```
+
+The object will contain a `<stream-id>` that defines the immutable content. We suffix the path with the version number to distinguish between different versions.
```

Comment:
Where exactly the object will contain the `<stream-id>`? In the object's content/body, or as a key in the UserMetadata?

Reply:
Object content since it's essential part of the structure. UserMetadata contains things that the satellite doesn't need.

Reply:
OK